Use real youtube thumbnails from youtube-dl

jgreco opened the pull request:
gallery-thumbgen.lua checks to see if the input_path starts with
https?://.  If so, it uses youtube-dl to attempt to resolve a url to a
thumbnail (i.e. the thumbnail used on the website.)  This works for any
site for which youtube-dl can find thumbnails.

In the special case of a link to youtube proper, youtube-dl is not used
because it's slow.  Instead the URL is parsed to pull out the video ID
and a URL to the thumbnail for that particular ID is generated.  This is
much faster than waiting for youtube-dl, but perhaps more brittle.

If any of the above fails, then the original input_path is returned and
thumbnailing proceeds as usual.

![screen shot 2018-12-07 at 4 31 58 pm](https://user-images.githubusercontent.com/41033/49679090-bdb03600-fa80-11e8-849f-2bf270c229e7.png)

occivink commented on scripts/gallery-thumbgen.lua:
```diff
@@ -1,8 +1,15 @@
+local ytdl = {
+    path = "youtube-dl",
+    searched = false,
+    blacklisted = {}
```
What's the purpose of this?

occivink commented on scripts/gallery-thumbgen.lua:
```diff
@@ -35,13 +42,72 @@ function is_video(input_path)
+    youtube_id1 = string.match(input_path, "https?://youtu.be/([%a%d%-_]+).*")
+    youtube_id2 = string.match(input_path, "https?://w?w?w?.youtube.com/v/([%a%d%-_]+).*")
+    youtube_id3 = string.match(input_path, "https?://w?w?w?.youtube.com/watch%?v=([%a%d%-_]+).*")
+    youtube_id4 = string.match(input_path, "https?://w?w?w?.youtube.com/embed/([%a%d%-_]+).*")
```
Make the `.` after www optional as well. Also escape them wherever necessary.

occivink commented on scripts/gallery-thumbgen.lua:
```diff
@@ -35,13 +42,72 @@ function is_video(input_path)
+    youtube_id2 = string.match(input_path, "https?://w?w?w?.youtube.com/v/([%a%d%-_]+).*")
+    youtube_id3 = string.match(input_path, "https?://w?w?w?.youtube.com/watch%?v=([%a%d%-_]+).*")
+    youtube_id4 = string.match(input_path, "https?://w?w?w?.youtube.com/embed/([%a%d%-_]+).*")
+    youtube_id = first_non_nil(youtube_id1, youtube_id2, youtube_id3, youtube_id4, false)
```
I think `youtube_id1 or youtube_id2 ...` should work

occivink commented on scripts/gallery-thumbgen.lua:
```diff
@@ -35,13 +42,72 @@ function is_video(input_path)
+            ytdl.path = ytdl_mcd
+        end
+        ytdl.searched = true
+    end
```
I'm not sure about this chunk, what's the use case for it? I think it's fine to just expect `youtube-dl` to be in your `$PATH`.

jgreco replied:
I pulled this 'searching for youtube-dl' code from ytdl_hook.lua a while ago.  My understanding is that many users (particularly windows users) won't have youtube-dl in their PATH but instead will have it sitting in their mpv config directory.   However looking at it again it appears I broke windows support in the process..

occivink replied:
Ah ok. From what I remember it was sufficient that ffmpeg/youtube-dl sits in the same folder as the mpv executable to be picked up by `subprocess` but if it's how ytdl_hook does it I suppose it's fine.

jgreco commented:
Fixed the pattern matching code, fixed finding youtube-dl on windows.

Also added ytdl blacklist code and an optional gallery_worker.conf to go with it.  (Personally I think this feature is superfluous, but ytdl_hook has it so maybe somebody somewhere has a good use for it...)
